### PR TITLE
Invalid charset

### PR DESCRIPTION
**Pull request by WhiteCat22**
This change makes Jaxrs requests with an invalid charset in the ```content-type``` header return HTTP 400 instead of 500 by throwing a ```BadRequestException``` instead of a generic ```IOException```.

However, this is only the fix for Jaxrs. After this change, the webcontainer throws it's own exception which causes the return code to still be 500 instead of 400 because they also process the ```content-type``` independently from Jaxrs.

Here is the new webcontainer stack:
```[3/20/20, 17:19:06:739 CDT] 00000045 com.ibm.ws.webcontainer.srt                                  I Unable to set request character encoding based upon request header 
java.io.UnsupportedEncodingException: SRVE0254E: Failed to set request character encoding: [invalid].
	at com.ibm.ws.webcontainer.srt.SRTServletRequest.setCharacterEncoding(SRTServletRequest.java:474)
	at com.ibm.ws.webcontainer.srt.SRTServletRequest.getCharacterEncoding(SRTServletRequest.java:1531)
	at com.ibm.ws.webcontainer.webapp.WebApp.sendError(WebApp.java:4164)
	at com.ibm.ws.webcontainer.webapp.WebApp.handleException(WebApp.java:5110)
	at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5090)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:314)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:1007)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1134)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:415)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:374)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:551)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:484)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:346)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:317)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:504)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:574)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:958)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1047)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:239)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:831)```

### CODE REVIEW

**Comment by WhiteCat22** [comment]
I disabled the test until webcontainer makes their change (if they decide to make the change)

**Comment by WhiteCat22** [comment]
#build

File: dev/com.ibm.ws.jaxrs.2.0_fat/fat/src/com/ibm/ws/jaxrs20/fat/helloworld/HelloWorldTest.java [hunk 1]
```diff
@@ -119,36 +126,18 @@ private StringBuilder runGetMethod(int exprc, String requestUri, String testOut)
         }
     }
 
-    private StringBuilder runDeleteMethod(int exprc, String testOut)
+    private int runGetMethodInvalidCharset(String requestUri)
```
Comment (WhiteCat22) [inline dev/com.ibm.ws.jaxrs.2.0_fat/fat/src/com/ibm/ws/jaxrs20/fat/helloworld/HelloWorldTest.java]: ignore the comparison, this is a new method that git wants to conflict with another one

**Comment by WhiteCat22** [comment]
#build pushed a new solution that didn't add a jaxrs api dependency to the transport.http bundle which would cause a hard dependency on jaxrs apis for jaxws-2.3.

File: dev/com.ibm.ws.jaxrs.2.0_fat/fat/src/com/ibm/ws/jaxrs20/fat/helloworld/HelloWorldTest.java (hunk 1 above)
Comment (andymc12) [reply dev/com.ibm.ws.jaxrs.2.0_fat/fat/src/com/ibm/ws/jaxrs20/fat/helloworld/HelloWorldTest.java]: Maybe you (or your IDE) deleted the old one?  I don't think anything calls it, so it's fine to remove it in any case.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/AbstractHTTPDestination.java [hunk 2]
```diff
@@ -36,6 +36,7 @@
 import javax.servlet.ServletContext;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+//import javax.ws.rs.BadRequestException;
```
[the remark this answers is not in the document]
Comment (WhiteCat22) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/AbstractHTTPDestination.java]: thought I did, oops

File: dev/com.ibm.ws.jaxrs.2.0_fat/fat/src/com/ibm/ws/jaxrs20/fat/helloworld/HelloWorldTest.java (hunk 1 above)
Comment (WhiteCat22) [reply dev/com.ibm.ws.jaxrs.2.0_fat/fat/src/com/ibm/ws/jaxrs20/fat/helloworld/HelloWorldTest.java]: ah, you are right, it was unused and I did delete it. Forgot about that.

**Comment by WhiteCat22** [comment]
#build after code review changes

**Comment by WhiteCat22** [comment]
#run-libby-bot

**Comment by pmd1nh** [comment]
WC does not throw any 500 for bad request encoding.  We logs and continues without any change to the status code.  I showed Adam my test's result.

**Pull request merged**